### PR TITLE
WIP: Add offset normalizer

### PR DESCRIPTION
**Pull request by phobson**
Trying to close https://github.com/matplotlib/matplotlib/issues/1806

Not currently on  a good machine for development, but I think this pretty straight-forward, so hopefully Travis a test runner will suffice...

`OffsetNorm`: This let's you set vmin, vcenter, and vmax. The use case for this is when you have data ranging from e.g., -1, to +2, and you  want the the white/gray part at 0, but would also like to use the full range of the red scale. So:

```python
offset = mcolors.OffsetNorm(vmin=-1, vcenter=0, vmax=2)
offset([-1.0, -0.5, 0.0, 1.0, 2.,])

#prints
[0, 0.25, 0.5, 0.75, 1.0]
```

Next I'll add `SymNorm`, that will look at the absolute values of vmin and vmax to produce:
```python
sym = mcolors.SymNorm(vmin=-1, vcenter=0, vmax=2)
sym([-1.0, -0.5, 0.0, 1.0, 2.,])

#prints
[0.25, 0.325, 0.5, 0.75, 1.0]
```

### CODE REVIEW

**Comment by phobson** [comment]
@tacaswell is this going into the next 1.4 or should I rebase on the 1.5.x branch?

**Comment by tacaswell** [comment]
This is aimed at 1.5.  It should be re-based onto current master please.

**Comment by phobson** [comment]
The log is too long to view on Travis directly, but the failure only occurs on the Python 3.4 test and appears unrelated. Here's the relevant part

```
======================================================================
FAIL: matplotlib.tests.test_mathtext.test_mathfont_stix_52.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/matplotlib-1.5.dev1-py3.4-linux-x86_64.egg/matplotlib/testing/decorators.py", line 51, in failer
    result = f(*args, **kwargs)
  File "/home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages/matplotlib-1.5.dev1-py3.4-linux-x86_64.egg/matplotlib/testing/decorators.py", line 196, in do_test
    '(RMS %(rms).3f)'%err)
matplotlib.testing.noseclasses.ImageComparisonFailure: images not close: /home/travis/build/matplotlib/tmp_test_dir/result_images/test_mathtext/mathfont_stix_52.png vs. /home/travis/build/matplotlib/tmp_test_dir/result_images/test_mathtext/mathfont_stix_52-expected.png (RMS 45.845)

----------------------------------------------------------------------
Ran 4915 tests in 116.972s

FAILED (KNOWNFAIL=17, SKIP=6, failures=1)
[Inferior 1 (process 10155) exited with code 01]
```

I'll restart the test machine and see if anything magic happens

**Comment by tacaswell** [comment]
Font rendering sometimes falls over.  I think we still have one or two tests that are missing '@cleanup' decorators and can stomp on each other.

**Comment by phobson** [comment]
hey @tacaswell  I think this is ready to go. I cleaned up and revised a few thing a while ago. I'll be unavailable to for the rest of the day today, but can look at things again tonight or tomorrow if anything still needs work.

**Comment by tacaswell** [comment]
At a minimum this will need an entry in 'whats_new`.

File: lib/matplotlib/colors.py [hunk 1]
```diff
@@ -969,6 +970,127 @@ def scaled(self):
...
+    """
+    def __init__(self, vmin=None, vcenter=None, vmax=None, clip=False):
+        """Normalize data with an offset midpoint
+
+        Useful when mapping data unequally centered around a conceptual
+        center, e.g., data that range from -2 to 4, with 0 as the midpoint.
+
+        Parameters
+        ----------
+        vmin : optional float
```
Comment (tacaswell) [inline lib/matplotlib/colors.py]: This should be 'float, optional' to conform to numpydoc style.

**Comment by phobson** [comment]
Took care of the docstring stuff and added a bit in `whats_new`. Will add an example when I'm on a machine that build MPL from source.

**Comment by phobson** [comment]
Looks like I've got some digging to do. I don't think this is how colorbars should be ticked with an `OffsetNorm`:

image

I was hoping that the ticks would be uniformly spaced along the cbar axis and -- in this case -- the blue portion of the cmap would simply transition to white that much faster.

**Comment by tacaswell** [comment]
Huh, makes sense why that happens (code path that makes log color scales work correctly).

**Comment by tacaswell** [comment]
Also see http://stackoverflow.com/questions/20144529/shifted-colorbar-matplotlib

I now no longer understand the color bar issue.

**Comment by phobson** [comment]
Yeah. I don't get it. I want to just close this and let Joe take over.

**Comment by tacaswell** [comment]
I the tests in here and the edge cases are worth keeping.  I think you just need to use `np.interp` instead of doing it by hand.

**Comment by phobson** [comment]
I'll take a crack this weekend. Thanks for the advice

**Comment by tacaswell** [comment]
@phobson The problem seems to be in the `inverse` method, remove that and it works the same.  The forward interpolation is correct, but we should use the numpy implementation.

**Comment by tacaswell** [comment]
attn @efiring this seems to go back to how the color bar is drawn which gitblame says you did a good bit of work on.

I think what is going on is that in most cases we want the color bar to be linear and the spacing between tick labels to be distorted, but in this case we want the tick labels to be linear and the color bar to be distorted.

**Comment by efiring** [comment]
@phobson, In your call to colorbar, try using the `spacing="proportional"` kwarg.

**Comment by OceanWolf** [comment]
Looking good, but quite a lot of tests fail with arrays not matching.

**Comment by phobson** [comment]
Yeah. I need to get on machine that can run the tests suite before I dive further into this. 



—

**Comment by OceanWolf** [comment]
Okay, but you can also just read off the errors from Travis.  Click on the details link, select a build (the first 2.7 build and the pep8 build should cover all the errors), then click on download log and search for ``FAIL`` or ``ERROR`` and you will see the traceback of the problem for normal builds and the coding style faux-pas in the pep8 build (see https://www.python.org/dev/peps/pep-0008/ for more information).

Running the tests on your own machine just means that you have to install ``nose``, but of course you can just copy-paste the code run it as a normal script, image checking you can do manually, but those might result from reasons relating to the other test failures, or a font problem in the image, but we can cross that bridge once we have no other errors.

**Comment by phobson** [comment]
@OceanWolf I know how to use Travis and Nose. I just don't have access to a machine where the test suite runs (the issue *isn't* nose), so pushing a bunch a trial and error commits to travis and waiting for the full test suite to run isn't really how I like to spend my weekend.

**Comment by phobson** [comment]
With the new `numpy.interp` approach, I wonder if it makes sense to allow people to arbitrary controls points. Something like:

```python
class OffsetNorm(Normalize):
    def __init__(self, *control_points, clip=False):
       self._controls_points = np.unique(sorted(control_points)).astype(float)
       self.vmin = min(self.self._controls_points)
       self.vmax = max(self._controls_points)

    def __call__(self, value, clip=None):
        if clip is None:
            clip = self.clip
        result, is_scalar = self.process_value(value)

        # yada yada 
        N = self._control_points.shape[0]
        x, y = self._control_points, np.linspace(0, 1, num=N, endpoint=True)
        return np.ma.masked_array(np.interp(value, x, y))
```

A bit of a self-imposed scope creep, I suppose, but this could be **very** handy with eg. baythmetric + topography data, radar reflectivity

**Comment by tacaswell** [comment]
I would vote for special casing the one point case and adding a second
general case.

We should also add an easy way to stack color maps to take advantage of the
control points.

I have seen examples of using color maps with sharp edges for medical
images (map the mri/cat value to the color the thing is in the body and
surgeons do better reading them).

**Comment by phobson** [comment]
Would you want that in a separate PR? 

On one hand (diff PR), we stay the course with this, but then have to refactor

On the other (this PR), we're pretty close to that point anyway and I don't mind sinking marginally more time into it once work settles down (this weekend then again early middle of next week).

**Comment by tacaswell** [comment]
I would like to get this in sooner rather than later, if it does not slow
you down one, else two

**Comment by joferkington** [comment]
@phobson - For what it's worth, another classic use case for "arbitrary control points" (more than just a midpoint and endpoints) is histogram equalization.  https://en.wikipedia.org/wiki/Histogram_equalization 

Implementing things that way (and making the midpoint a convenient subclass) would allow one to easily add a ``HistogramEqualizatonNorm`` with just a few lines of code.

Also, here's another use-case for the arbitrary control points from an SO question from a few weeks ago: http://stackoverflow.com/q/30465080/325565  (My way-too-short answer there is actually fraught with difficulties in the details because I didn't bother to define the inverse, etc, that you've done quite nicely.)

**Comment by phobson** [comment]
@joferkington funny than you mention the `inverse` method. I found it was actually that method that was making the tick marks spaced so crazy on the colorbar (see the inline image up thread). I looked into the `LogNorm` class and it didn't have `inverse` defined. So I removed it from this PR (see the PR's baseline image) and that fixed the ticks.

Point is, I'm curious if I actually do need an `inverse` method and had simply gone about it the wrong way. If not, then your latest SO solution might be closer than you think.

On *that* note, I was going to take almost an identical approach as you did. Only tweak would be that `cvalues` defaults to `None` in the signature and gets sets to `np.linspace(0, 1, num=len(xvalues), endpoint=True)`. Seem reasonable?

**Comment by tacaswell** [comment]
@phobson Did you try the colorbar kwarg suggested by eric?

**Comment by phobson** [comment]
I did. It didn't change anything, though I was working on a janky dev installation of MPL, so I should probably give it another shot. 

Looking back a existing source code, `LogNorm` *does* have an `inverse` method, so there's something to learn there I'm sure.

**Comment by OceanWolf** [comment]
Just had an idea in terms of a name for the general case, ``GraduatedNorm``, what do you think?

**Comment by efiring** [comment]
It sounds like all this is actually about using a continuous piece-wise linear curve (first-degree spline) as the mapping function.  Is that correct?  `GraduatedNorm` doesn't give me any picture of what it is supposed to be.

**Comment by Tillsten** [comment]
`PiecewiseLinearNorm` sounds good to me.

**Comment by WeatherGod** [comment]
I like that name.

**Comment by OceanWolf** [comment]
:+1:

**Comment by tacaswell** [comment]
Closing in favor of #4666 @phobson ping to have this re-opened if you want to take over development again.
